### PR TITLE
Shared Memory tests using Python Client API

### PR DESCRIPTION
**Pull request by CoderHam**
- [x] - L0_infer
- [x] - L0_batcher
- [x] - L0_sequence_batcher
- [x] L0_shared_memory - test shared memory client API functionality

### CODE REVIEW

File: qa/L0_batcher/batcher_test.py [hunk 1]
```diff
@@ -151,6 +152,26 @@ def test_static_batch_preferred(self):
             except InferenceServerException as ex:
                 self.assertTrue(False, "unexpected error {}".format(ex))
 
+    def test_static_batch_preferred_shm(self):
```
Comment (GuanLuo) [inline qa/L0_batcher/batcher_test.py]: Only this case is tested with shared memory?

File: qa/common/infer_util.py [hunk 2]
```diff
@@ -55,15 +57,21 @@ def infer_exact(tester, pf, tensor_shape, batch_size,
...
                 skip_request_id_check=False, use_streaming=True,
-                correlation_id=0):
+                correlation_id=0, use_shared_memory=False):
     tester.assertTrue(use_http or use_grpc or use_streaming)
     configs = []
     if use_http:
-        configs.append(("localhost:8000", ProtocolType.HTTP, False))
+        if use_shared_memory:
+            configs.append(("localhost:8000", ProtocolType.HTTP, False, True))
+        configs.append(("localhost:8000", ProtocolType.HTTP, False, False))
```
Comment (GuanLuo) [inline qa/common/infer_util.py]: So if `use_shared_memory == True`, the "no shared memory" case will still be tested?

File: qa/common/infer_util.py [hunk 3]
```diff
@@ -143,24 +170,38 @@ def infer_exact(tester, pf, tensor_shape, batch_size,
...
-                output_req[OUTPUT0] = InferContext.ResultFormat.RAW
-            else:
-                output_req[OUTPUT0] = (InferContext.ResultFormat.CLASS, num_classes)
-        if "OUTPUT1" in outputs:
-            if output1_raw:
-                output_req[OUTPUT1] = InferContext.ResultFormat.RAW
-            else:
-                output_req[OUTPUT1] = (InferContext.ResultFormat.CLASS, num_classes)
+        if config[3]:
+            if "OUTPUT0" in outputs:
```
Comment (GuanLuo) [inline qa/common/infer_util.py]: change the order:
```
if "OUTPUT0" in outputs:
    if config[3]:
        ...
    elif output0_raw:
        ...
    else:
        ...
```

File: src/clients/python/crequest.cc [hunk 4]
```diff
@@ -29,6 +29,7 @@
 #include <iostream>
 #include "src/clients/c++/request_grpc.h"
 #include "src/clients/c++/request_http.h"
+#include "src/clients/python/shared_memory/shared_memory_handle.h"
```
Comment (GuanLuo) [inline src/clients/python/crequest.cc]: Isn't this the wrapper for the Python only? The request API doesn't need to know anything about it, right?

File: qa/common/infer_util.py (hunk 2 above)
Comment (deadeyegoodwin) [reply qa/common/infer_util.py]: include_shared_memory_test might be a better name to imply that in addition to regular testing also want to do shared-memory testing.

**Comment by deadeyegoodwin** [comment]
@CoderHam Need to rebase this to master to get rid of the already committed changes.

File: src/clients/python/crequest.cc (hunk 4 above)
Comment (CoderHam) [reply src/clients/python/crequest.cc]: This contains the shared_memory_handle struct used to get the data inside it from the void ptr handle that was returned to python when the shared memory region was created.

File: qa/common/infer_util.py (hunk 2 above)
Comment (CoderHam) [reply qa/common/infer_util.py]: Done

File: qa/common/infer_util.py (hunk 3 above)
Comment (CoderHam) [reply qa/common/infer_util.py]: Sure

File: qa/L0_batcher/batcher_test.py (hunk 1 above)
Comment (CoderHam) [reply qa/L0_batcher/batcher_test.py]: Modified for greater test coverage.

**Pull request merged**